webui/react/src/components/FilterForm/components/FilterFormStore.ts

```diff
@@ -0,0 +1,191 @@
+  value: null,
+});
+
+export class FilterFormStore {
+  #formset: WritableObservable<FilterFormSet>;
+
+  constructor(data?: Readonly<FilterFormSet>) {
+    this.init(data);
+  }
+
+  public init(data?: Readonly<FilterFormSet>): void {
+    this.#formset = observable<FilterFormSet>(structuredClone(data ? data : INIT_FORMSET));
+  }
+
+  public get formset(): Observable<Readonly<FilterFormSet>> {
+    return this.#formset.readOnly();
+  }
+
+  public get jsonWithoutId(): Observable<Readonly<FilterFormSetWithoutId>> {
+    const replacer = (key: string, value: unknown): unknown => {
+      return key === 'id' ? undefined : value;
+    };
+    return this.#formset.select((formset) => {
+      const newFormSet: FilterFormSetWithoutId = JSON.parse(JSON.stringify(formset, replacer));
+      return newFormSet;
+    });
+  }
+
+  #getFormById(filterGroup: FormGroup, id: string): FormField | FormGroup | undefined {
+    const recur = (form: FormGroup | FormField): FormGroup | FormField | undefined => {
+      if (form.id === id) {
+        return form;
+      }
+      if (form.kind === FormKind.Group && form.children.length === 0) {
+        return undefined;
+      }
+
+      if (form.kind === FormKind.Group) {
+        for (const child of form.children) {
+          const ans = recur(child);
+          if (ans) {
+            return ans;
+          }
+        }
+      }
+      return undefined;
+    };
+
+    return recur(filterGroup);
+  }
+
+  public setFieldColumnName(
+    id: string,
+    col: Pick<V1ProjectColumn, 'location' | 'type' | 'column'>,
+  ): void {
+    const filterSet: Readonly<FilterFormSet> = this.#formset.get();
+    const filterGroup = filterSet.filterGroup;
+    const ans = this.#getFormById(filterGroup, id);
+    if (ans && ans.kind === FormKind.Field) {
+      ans.columnName = col.column;
+      ans.location = col.location;
+      ans.type = col.type;
+      this.#formset.set({ filterGroup, showArchived: filterSet.showArchived });
+    }
+  }
+
+  public setFieldOperator(id: string, operator: Operator): void {
+    const filterSet: Readonly<FilterFormSet> = this.#formset.get();
+    const filterGroup = filterSet.filterGroup;
+    const ans = this.#getFormById(filterGroup, id);
+    if (ans && ans.kind === FormKind.Field && Object.values(Operator).includes(operator)) {
+      ans.operator = operator;
+      this.#formset.set({ filterGroup, showArchived: filterSet.showArchived });
+    }
+  }
+
+  public setFieldConjunction(id: string, conjunction: Conjunction): void {
+    const filterSet: Readonly<FilterFormSet> = this.#formset.get();
+    const filterGroup = filterSet.filterGroup;
+    const ans = this.#getFormById(filterGroup, id);
+    if (ans && ans.kind === FormKind.Group && Object.values(Conjunction).includes(conjunction)) {
+      ans.conjunction = conjunction;
+      this.#formset.set({ filterGroup, showArchived: filterSet.showArchived });
+    }
+  }
+
+  public setFieldValue(id: string, value: FormFieldValue): void {
```

Comment:
you should be able to merge `setFieldValue`, `setFieldColumnName` and `setFieldOperator` into a single function that takes a partial update, no?

Reply:
it used to a single function, but i didnt look good, so i split into different functions